### ISSUE TITLE
Wrong number of cosmics for small images

### ISSUE THREAD

**Issue by dk766** [IvS-KULeuven/PlatoSim3#283]
Platosim version used: PlatoSim v3.2.0-RC1-482-g9e0ff1b
Main config params:
Image size: 10x10
Stars: 1 (mag 30)
Cosmic rates: 10
Number of exposures 20 000

Total number of outliers: 0
Considering the large number of exposures generated the total number of expected cosmics  is around 1360. 

pixel surface x 100 pixels x 21 seconds x 20 000 exposures = 1360

I attached the config files used (renamed inputfile.yaml to inputfile.txt)
[starcatalog_onestar.txt](https://github.com/IvS-KULeuven/PlatoSim3/files/2374791/starcatalog_onestar.txt)
[inputfile.txt](https://github.com/IvS-KULeuven/PlatoSim3/files/2374795/inputfile.txt)

**Comment by dk766** [comment]
I re-run the simulation with an increased number of cosmics: 100 /cm2/s and I obtained in total 214 cosmics which is also too small.

An additional request when solving this issue: could you please change the log message when reporting the number of cosmics that affect different CCD parts ? 

Currently for all cosmics affecting subfield, smearing map, bias map the message is exactly similar: 

> number of cosmic hits: N

which makes it difficult to count a specific type of cosmic. You could maybe have: 

> number of cosmic hits in bias map: N 
> number of cosmic hits in smearing map: M


which would make grep-ing the log much easier.

**Comment by JorisDeRidder** [comment]
Thanks for the report. We're working on it.

**Comment by SaraRegibo** [comment]
As a start, I improved the log messages as in the following example:

```
2018-09-12 15:02:46 DEBUG   Detector: including cosmic hits.
2018-09-12 15:02:46 DEBUG   Detector: adding cosmic hits to the sub-field
2018-09-12 15:02:46 DEBUG   Detector: number of cosmic hits for the image area: 9
2018-09-12 15:02:46 INFO    Detector: mean cosmic entry angle (image area): 2.794298 rad
2018-09-12 15:02:46 INFO    Detector: mean cosmic trail length (image area): 7.289248 pix
2018-09-12 15:02:46 INFO    Detector: mean cosmic hit intensity (image area): 23024.706993 e-
2018-09-12 15:02:46 DEBUG   Detector: adding cosmic hits to smearing map
2018-09-12 15:02:46 DEBUG   Detector: number of cosmic hits for the smearing map: 0
2018-09-12 15:02:46 DEBUG   Detector: adding cosmic hits to bias map
2018-09-12 15:02:46 DEBUG   Detector: number of cosmic hits for the bias map: 0
```

**Comment by SaraRegibo** [comment]
For each exposure, we follow the following procedure:
* The cosmic hit rate is not the same for all exposure but is sampled from a Poisson distribution with the `CosmicHitRate` from the configuration file as mean.  So the first step is to determine the cosmic hit rate for the current exposure by drawing from this exposure.  This is taken care of by the term `cosmicHitRateDistribution(cosmicHitRateGenerator)` in the formula below.
* The next step is to determine the actual number of cosmic hits in the sub-field for the current exposure, which is done as follows:

```
int numCosmicHits = cosmicHitRateDistribution(cosmicHitRateGenerator) * exposureTime * (numRows * pixelSize / 10000.0) * (numColumns * pixelSize / 10000.0);
```

For the case above - without casting to `int` - this will be of the order (cosmic hit rate: 10 events / cm^2 / s):

```
cosmicHitRate * pixelArea * exposureTime
Out[15]: 0.07128
```

But the  number of cosmics should be an integer, so the number above will be rounded to zero, meaning: no cosmics in the sub-field for the current exposure.

**Comment by JorisDeRidder** [comment]
A value of 0.07128 hits per single exposure can be interpreted as ~7 hits per 100 exposures (on average). So for a number numHits < 1.0, I would recommend to generate a uniform number u between 0 and 1. If u < numHits then numHits = 1, otherwise numHits = 0. Such a procedure guarantees the correct average of cosmic hits, as well as their random nature.

**Comment by reneheller** [comment]
I understand that 0.07128 is the expectation value for a single exposure, which is equivalent to the arithmetic mean for an infinitely large number of realizations. For any given realization in PLATOSim, as we have it, the actual value could be 0 cosmics (very likely), 1 cosmic (much less likely), 2 cosmics (even less likely), etc. But it can't just always be 0 or otherwise one would never obtain an average of 0.07128.

**Comment by JorisDeRidder** [comment]
As a follow-up, the correct procedure is probably the following. Generate a uniform random number u between 0 and 1. Let f be the fractional part of the number of hits (e.g. numHits = 0.07 -> f = 0.07,  numHits = 2.45 -> f = 0.45). If  u < f: round up numHits, otherwise round down. This ensures that 2.45 will not always be rounded down so that the number of hits is systematically underestimated. In 45% of the cases it will be round up to 3.0, so that the average hit rates is indeed 2.45. For large hit rates, this will not make a lot of difference, but for small hit rates it will.

**Comment by dk766** [comment]
Just one note to this alg.: pls. make sure that "u" is generated using the cosmics random number generator (or if not possible one for which the seed is configurable in the inputfile.yaml) to ensure that by setting the same seed I obtain the same realization of cosmics.

**Comment by SaraRegibo** [comment]
Finally some time to work on this!  I will implement it as @JorisDeRidder suggested above.  All random generators for the cosmics (i.e. cosmic hit rate, entry row and column, position angle,...) already use a seed derived from the same configuration parameter (RandomSeeds: CosmicSeed).  I will make sure to do this for the new random generator, too.   This will ensure the results to be reproducible, as asked by @dk766 .

**Comment by SaraRegibo** [comment]
Solved in the `develop` branch in #286 .  @dk766 could you check and close this issue if it works (or re-assign it to me if it doesn't)?

If you also want me to patch it in `master`, please, re-assign the issue back to me.  Thanks.

**Comment by dk766** [comment]
I rerun the simulations and I confirm that the current behaviour is the expected one. Pls. also merge into master

**Comment by SaraRegibo** [comment]
* `develop`: solved by #286 
* `master`: solved by #287 (release 3.3.4)
